src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala

```diff
@@ -178,24 +181,50 @@ final case class AgreementApiServiceImpl(
     val operationLabel = s"Upgrading agreement $agreementId"
     logger.info(operationLabel)
 
-    val result: Future[Agreement] = for {
-      requesterOrgId <- getOrganizationIdFutureUUID(contexts)
-      agreementUUID  <- agreementId.toFutureUUID
-      agreement      <- agreementManagementService.getAgreementById(agreementUUID)
-      _              <- assertRequesterIsConsumer(requesterOrgId, agreement)
-      _              <- agreement.assertUpgradableState.toFuture
-      eService       <- catalogManagementService.getEServiceById(agreement.eserviceId)
-      newDescriptor  <- CatalogManagementService.getEServiceNewerPublishedDescriptor(eService, agreement.descriptorId)
-      uid            <- getUidFutureUUID(contexts)
-      stamp       = Stamp(uid, offsetDateTimeSupplier.get())
+    def upgradeAgreement(
+      oldAgreementId: UUID,
+      newDescriptor: EServiceDescriptor
+    ): Future[AgreementManagement.Agreement] = for {
+      stamp <- getUidFutureUUID(contexts).map(Stamp(_, offsetDateTimeSupplier.get()))
       upgradeSeed = AgreementManagement.UpgradeAgreementSeed(descriptorId = newDescriptor.id, stamp)
-      newAgreement <- agreementManagementService.upgradeById(agreement.id, upgradeSeed)
+      newAgreement <- agreementManagementService.upgradeById(oldAgreementId, upgradeSeed)
       payload = getClientUpgradePayload(newAgreement, newDescriptor)
       _ <- authorizationManagementService.updateAgreementAndEServiceStates(
         newAgreement.eserviceId,
         newAgreement.consumerId,
         payload
       )
+    } yield newAgreement
+
+    def createNewDraftAgreement(agreement: AgreementManagement.Agreement, newDescriptorId: UUID) =
+      for {
+        _         <- verifyConflictingAgreements(agreement, AgreementManagement.AgreementState.DRAFT :: Nil)
+        agreement <- agreementManagementService.createAgreement(
+          AgreementManagement.AgreementSeed(
+            eserviceId = agreement.eserviceId,
+            descriptorId = newDescriptorId,
+            producerId = agreement.producerId,
+            consumerId = agreement.consumerId,
+            verifiedAttributes = agreement.verifiedAttributes.map(a => AgreementManagement.AttributeSeed(a.id)),
+            certifiedAttributes = agreement.certifiedAttributes.map(a => AgreementManagement.AttributeSeed(a.id)),
+            declaredAttributes = agreement.declaredAttributes.map(a => AgreementManagement.AttributeSeed(a.id)),
+            consumerNotes = agreement.consumerNotes
+          )
+        )
+      } yield agreement
```

Comment:
The `consumerDocuments` should be copied too.
You can use the function `createAndCopyDocumentsForClonedAgreement` (see the comment below).

Reply:
good catch

Reply:
I would rather prefer to make it "more" atomic, and modify the AgreementSeed in both management and process, wdyt?

Reply:
The new field will be passed nonempty only in this specific case.